docs/testing.md

```diff
@@ -213,7 +240,96 @@ public void shouldSubstractAfterAdd() {
+> with `Test`.
+
+Of course, sometimes you're not testing a class, but rather multiple as one unit. In this case,
+name the test class such that it conveys the logical grouping of the tests it contains.
+
+For example, look at
+the [ObjectMappingDefaultValuesTest](zeebe/msgpack-value/src/test/java/io/camunda/zeebe/msgpack/ObjectMappingDefaultValuesTest.java),
+which tests the serialization and deserialization of default properties in `UnpackedObject` types.
+Or
+the [JsonSerializableToJsonTest](zeebe/protocol-impl/src/test/java/io/camunda/zeebe/protocol/impl/JsonSerializableToJsonTest.java),
+which tests that all protocol values can be serialized to JSON.
+
+In both cases, we are testing a few classes interacting together, but testing logically similar
+behavior (e.g. JSON serialization) which warrants grouping together in the same test class.
+
+Note that these classes still live in the same modules as the units they are testing, even if they
+group multiple ones together.
+
+### When to write a unit test
+
+As a rule of thumb, you should always add unit tests for any public API with some business logic.
+Avoid testing very simple methods (e.g. getter, setter), and focus on those with business logic.
+
+Additionally, focus on public facing APIs only. Avoid testing private, protected, or package private
+methods as these are implementation details: if something is not accessible from outside the unit,
+it's essentially an implementation detail. By focusing on testing only the public behavior, we avoid
+coupling the test to the implementation, making it more resilient to future changes.
+
+> [!Note]
+> Of course, if a class is package-private and used extensively in its own package, its public API
+> is then its private-package level methods. Use your judgement on what is the public API of a unit.
+
+Finally, bug fixes should always come with at least one regression test.
+
+> [!Note]
+> You can use the
+> [RegressionTest](zeebe/test-util/src/main/java/io/camunda/zeebe/test/util/junit/RegressionTest.java)
+> annotation where you can add a link to issue being fixed. This provides context for future
+> maintainers as to why the test is here at all.
+
+Let's illustrate this with some examples. Take one of our existing
+classes, [ClientStreamImpl](zeebe/transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamImpl.java):
+
+```java
+/** Represents a registered client stream. * */
+record ClientStreamImpl<M extends BufferWriter>(
+    ClientStreamIdImpl streamId,
+    AggregatedClientStream<M> serverStream,
+    DirectBuffer streamType,
+    M metadata,
+    ClientStreamConsumer clientStreamConsumer)
+  implements ClientStream<M> {
+
+  @Override
+  public Set<MemberId> liveConnections() {
+    return serverStream().liveConnections();
+  }
+}
+```
+
+It would be pointless to test that the properties are returned by the getters; not only is this
+logic too simple, but it's also essentially testing a Java feature, not our own code.
```

Comment:
Java or is it a grpc feature?

Reply:
These are our own stream between gateways and brokers :) We don't have gRPC between. So they're just plain vanilla Java.